### PR TITLE
feat(medusa,medusa-react): PaymentCollection support

### PR DESCRIPTION
**Pull request by carlos-r-l-rodrigues**
what:

medusa-react implementation of payment collections and payments
endpoint to handle a single payment session of Payment Collections
endpoint to handle multiple payment sessions of Payment Collections

FIXES CORE-880, CORE-881, CORE-882

### CODE REVIEW

File: packages/medusa-react/src/hooks/admin/payment-collections/index.ts
Comment (olivermrbl) [inline packages/medusa-react/src/hooks/admin/payment-collections/index.ts:2]: todo: Should use double quotes `"`
```suggestion
export * from "./mutations"
```

File: packages/medusa-react/src/hooks/admin/payment-collections/queries.ts
Comment (olivermrbl) [inline packages/medusa-react/src/hooks/admin/payment-collections/queries.ts:16]: todo: Missing domain from name
```suggestion
export const useAdminPaymentCollection = (
```

File: packages/medusa-react/src/hooks/admin/payments/index.ts
Comment (olivermrbl) [inline packages/medusa-react/src/hooks/admin/payments/index.ts:2]: todo: Same as other comment.
```suggestion
export * from "./mutations"
```

File: packages/medusa-react/src/hooks/admin/payment-collections/queries.ts
Comment (olivermrbl) [inline packages/medusa-react/src/hooks/admin/payment-collections/queries.ts:19]: todo: Should use admin payment collection response

File: packages/medusa-react/src/hooks/admin/payments/queries.ts
Comment (olivermrbl) [inline packages/medusa-react/src/hooks/admin/payments/queries.ts:15]: todo: Missing domain from name
```suggestion
export const useAdminPayment = (
```

File: packages/medusa-react/src/hooks/store/payment-collections/mutations.ts
Comment (olivermrbl) [inline packages/medusa-react/src/hooks/store/payment-collections/mutations.ts:39]: suggestion(non-blocking): Rename to remove ambiguity
```suggestion
export const useAuthorizePayment[Session | Collection] = (
```

Comment (olivermrbl) [reply packages/medusa-react/src/hooks/store/payment-collections/mutations.ts:39]: Not entirely sure, if we should use session or collection here. Collection is primarily a backend concept, which is why session might make sense. What do you think?

File: packages/medusa-react/src/hooks/store/payment-collections/index.ts
Comment (olivermrbl) [inline packages/medusa-react/src/hooks/store/payment-collections/index.ts:2]: todo: Same as other comment
```suggestion
export * from "./mutations"
```

**Review by olivermrbl — CHANGES_REQUESTED** [review]
Nice work. Have left a couple of todos 💪 

We seem to be missing tests. If you want to have that in a separate PR, that's fine with me.

File: packages/medusa-react/src/hooks/store/payment-collections/mutations.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa-react/src/hooks/store/payment-collections/mutations.ts:39]: The payCol authorization will try to authorize all payment sessions that are not yet authorized.
Do you think we should have an authorizeSession method that authorizes session by session?

File: packages/medusa/src/api/routes/store/payment-collections/index.ts
Comment (srindom) [inline packages/medusa/src/api/routes/store/payment-collections/index.ts:61]: If you change the order of the batch operations they will work. See this example: https://codesandbox.io/s/express-js-forked-ci0l54?file=/src/index.js

I.e. make sure that:

`.post("/:id/sessions/batch/authorize" ...` is defined before `.post("/:id/sessions/:session_id/authorize" ...`

Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/api/routes/store/payment-collections/index.ts]: done. 😄 nobody will be able to use `batch` as a payment session id.

File: packages/medusa/src/services/payment-collection.ts
Comment (srindom) [inline packages/medusa/src/services/payment-collection.ts:304]: nit: should be `customerId`

File: packages/medusa/src/services/payment-collection.ts
Comment (srindom) [inline packages/medusa/src/services/payment-collection.ts:347]: thought: is this necessary?  since there already is a resource id 6 lines above?

File: packages/medusa/src/services/payment-collection.ts
Comment (srindom) [inline packages/medusa/src/services/payment-collection.ts:396]:
```suggestion
    customerId: string
```

File: packages/medusa-js/src/resources/payment-collections.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa-js/src/resources/payment-collections.ts:56]: It is also updating and deleting them. I don't know if create would be the best name for it.

File: packages/medusa-js/src/resources/payment-collections.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa-js/src/resources/payment-collections.ts:65]: It means remove the current session and create a new one. I followed what is done for the cart, so I assumed it is some edge case where old sessions need to be recreated. If that's not the case, theis method could be removed

File: packages/medusa/src/api/middlewares/authenticate-customer.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/api/middlewares/authenticate-customer.ts:17]: To avoid assign `false` to it.
It should be either undefined or the object

File: packages/medusa/src/api/middlewares/require-customer-authentication.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/api/middlewares/require-customer-authentication.ts:6]: It this exist, we can skip ` passport.authenticate` that was called previously in another middleware.

File: packages/medusa/src/api/routes/store/order-edits/index.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/api/routes/store/order-edits/index.ts:27]: Should we let all the endpoint of order edit open?
I let the payment collection ones

File: packages/medusa/src/api/routes/store/payment-collections/authorize-payment-session.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/api/routes/store/payment-collections/authorize-payment-session.ts:76]: Will change that then. Any suggestion of Error Type for it?

File: packages/medusa/src/services/payment-collection.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/payment-collection.ts:347]: Will check that. Pretty sure that is being used by a payment plugin, but maybe that can be adapted there.

File: packages/medusa/src/api/routes/store/order-edits/index.ts
Comment (srindom) [reply packages/medusa/src/api/routes/store/order-edits/index.ts:27]: Yes everything should be accessible without having to log in :)

File: packages/medusa/src/api/routes/store/payment-collections/authorize-payment-session.ts
Comment (srindom) [reply packages/medusa/src/api/routes/store/payment-collections/authorize-payment-session.ts:76]: Think it would be okay to introduce a payment specific error like:  `MedusaError.Types.PAYMENT_AUTHORIZATION_ERROR`

File: packages/medusa/src/api/middlewares/require-customer-authentication.ts
Comment (srindom) [reply packages/medusa/src/api/middlewares/require-customer-authentication.ts:6]: Makes sense thanks for clarifying :)

File: packages/medusa-js/src/resources/payment-collections.ts
Comment (srindom) [reply packages/medusa-js/src/resources/payment-collections.ts:56]: Got it -- fine to keep as is

File: packages/medusa/src/services/payment-collection.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/payment-collection.ts:347]: nice catch.. Thanks.
I mixed a few concepts. This metadata is not needed.

File: packages/medusa/src/services/payment-collection.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/payment-collection.ts:257]: added JSDocs to payment collection and payment services

**Review by srindom — APPROVED** [review]
LGTM!

Pending @olivermrbl's approval :)

File: packages/medusa/src/services/payment-collection.ts
Comment (olivermrbl) [inline packages/medusa/src/services/payment-collection.ts:378]: thought: In the Stripe payment plugin, we have the following check when updating a payment session:
```
      const stripeId = paymentInput.customer?.metadata?.stripe_id

      if (stripeId !== paymentInput.customer_id) {
        return await this.createPaymentNew(paymentInput, intentRequestData)
      } else {
        if (paymentSessionData.amount === Math.round(paymentInput.amount)) {
          return paymentSessionData
        }

        return await this.stripe_.paymentIntents.update(paymentSessionData.id, {
          amount: Math.round(paymentInput.amount),
        })
      }
```

But with this new implementation, I don't think we ever add `customer_id` in the flow from the payment collection service to the payment provider plugin service. Shouldn't we add that alongside the customer to make the logic in the Stripe provider kick in properly?

**Review by olivermrbl — COMMENTED** [review]
Great work @carlos-r-l-rodrigues. Just have a couple of comments, that I'd like you to address before we merge :)

File: integration-tests/api/__tests__/store/order-edit.js
Comment (carlos-r-l-rodrigues) [reply integration-tests/api/__tests__/store/order-edit.js:26]: will do

File: packages/medusa/src/api/routes/store/payment-collections/manage-payment-session.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/api/routes/store/payment-collections/manage-payment-session.ts:7]: If we send the provider_id at the URL I believe it won't be clear that only the last provider_id will be selected and other ones will be deleted/canceled. Or is the idea to keep the flow like carts where all payment sessions are created and the endpoint just set the flag `selected` in one of them?

People might think they can control multiple sessions with multiple providers by calling this endpoint using different provider_ids.
Maybe we can rename `/sessions` but I believe the provider_id should be passed as the payload.
But I don't have a strong opinion on it. I am ok with other approaches.

File: packages/medusa/src/services/payment-collection.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/payment-collection.ts:87]: findOne executes 2 queries, first one to count the number of rows and then it query again to return the data. This is done internally by TypeORM in case there are relations returning multiple rows and it has to be aggregated to the main object.
I wanted to avoid this

File: packages/medusa/src/api/routes/store/payment-collections/manage-payment-session.ts
Comment (olivermrbl) [reply packages/medusa/src/api/routes/store/payment-collections/manage-payment-session.ts:7]: > Or is the idea to keep the flow like carts where all payment sessions are created and the endpoint just sets the flag selected in one of them

Good question. I believe it's a tradeoff. Keeping the sessions would lead to a more performant endpoint when the customer is shuffling between the session, but deleting/canceling them would clean up the DB and eventually free up space.

I am fine keeping it as is, and then we can reevaluate based on feedback from users.

File: packages/medusa/src/api/routes/store/payment-collections/authorize-payment-session.ts
Comment (olivermrbl) [reply packages/medusa/src/api/routes/store/payment-collections/authorize-payment-session.ts:74]: Sorry, my bad :D

File: packages/medusa/src/services/payment-collection.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/payment-collection.ts:378]: Thanks for seeing this. That actually uncovered a bigger issue when the customer is empty. customerService.retrieve is fetching a customer when the customerId is undefined and currently stripe plugin is requiring at least a customer email to work.
I will address these issues.

Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/payment-collection.ts:382]: For that PR I'm checking the customerId undefined before calling customerService.retrive, but I will try to handle these undefined ids in all retrieve methods in another PR.
This can cause some weird hidden issues in multiple places

Comment (olivermrbl) [reply packages/medusa/src/services/payment-collection.ts:382]: Great - sounds good!

**Review by olivermrbl — APPROVED** [review]
LGTM!

**Pull request merged**
